### ISSUE TITLE
Referenced but unset environment variable evaluates to an empty string: MAINPID

### ISSUE THREAD

**Issue by Jappie3** [Infinidoge/nix-minecraft#119]
systemd service doesn't start:

```
Dec 28 22:16:58 Flugel systemd[1]: Starting Minecraft Server smp...
Dec 28 22:16:59 Flugel systemd[1]: Started Minecraft Server smp.
Dec 28 22:17:00 Flugel (smp-stop)[193304]: minecraft-server-smp.service: Referenced but unset environment variable evaluates to an empty string: MAINPID
Dec 28 22:17:00 Flugel minecraft-server-smp-stop[193306]: no server running on /run/minecraft/smp.sock
Dec 28 22:17:00 Flugel systemd[1]: minecraft-server-smp.service: Deactivated successfully.
Dec 28 22:17:00 Flugel systemd[1]: minecraft-server-smp.service: Consumed 2.169s CPU time, 130M memory peak, 4K written to disk.
Dec 28 22:17:00 Flugel systemd[1]: minecraft-server-smp.service: Scheduled restart job, restart counter is at 2.
```

service file:

```
$ systemctl cat minecraft-server-smp
# /etc/systemd/system/minecraft-server-smp.service
[Unit]
After=network.target
Description=Minecraft Server smp
StartLimitBurst=5
StartLimitIntervalSec=120

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/gzali8g9fnxfzd2i15g0xmvzywx24hib-glibc-locales-2.40-36/lib/locale/locale-archive"
Environment="PATH=/nix/store/b1wvkjx96i3s7wblz38ya0zr8i93zbc5-coreutils-9.5/bin:/nix/store/w8pnfazxqwmrqmwkb5zrz1bifsd8abxl-findutils-4.10.0/bin:/nix/store/85amyk92rg19l4fy0qmy7wr4jmq8p5z0-gnugrep-3.11/bin:/nix>
Environment="TZDIR=/nix/store/dfsiy184c8z6gf5x6cqfyahxv9g16828-tzdata-2024b/share/zoneinfo"
CapabilityBoundingSet=
DeviceAllow=
ExecReload=/nix/store/3anqy4q4ddcc65r3gss83jr66dyrmydh-minecraft-server-smp-reload/bin/minecraft-server-smp-reload
ExecStart=/nix/store/71vflga6nbfhsi318xa83xc51b14nhds-minecraft-server-smp-start/bin/minecraft-server-smp-start
ExecStartPost=/nix/store/8rlmv1rm51wlahkl4f9xn7jra6kij5g4-minecraft-server-smp-start-post/bin/minecraft-server-smp-start-post
ExecStartPre=/nix/store/858ppq08q9fg0gcbgfdb75ksx38azk9k-minecraft-server-smp-start-pre/bin/minecraft-server-smp-start-pre
ExecStop=/nix/store/b67skmhr9bln9ihksbw2qgpf0grbhdw7-minecraft-server-smp-stop/bin/minecraft-server-smp-stop $MAINPID
ExecStopPost=/nix/store/cgmb5g6xipng9sf12qxm2z6migcmmg7v-minecraft-server-smp-stop-post/bin/minecraft-server-smp-stop-post
Group=minecraft
GuessMainPID=true
LockPersonality=true
PrivateDevices=true
PrivateTmp=true
PrivateUsers=true
ProtectClock=true
ProtectControlGroups=true
ProtectHome=true
ProtectHostname=true
ProtectKernelLogs=true
ProtectKernelModules=true
ProtectKernelTunables=true
ProtectProc=invisible
Restart=always
RestrictAddressFamilies=AF_UNIX
RestrictAddressFamilies=AF_INET
RestrictAddressFamilies=AF_INET6
RestrictNamespaces=true
RestrictRealtime=true
RestrictSUIDSGID=true
RuntimeDirectory=minecraft
RuntimeDirectoryPreserve=yes
SystemCallArchitectures=native
TimeoutStopSec=1min 15s
Type=forking
UMask=0007
User=minecraft
WorkingDirectory=/var/lib/minecraft/smp

[Install]
WantedBy=multi-user.target
```

I also don't really understand why `MAINPID` is passed as an argument to `execStopScript` (`minecraft-server-<name>-stop`) as it doesn't seem like the script uses any variables...

<details><summary>service config</summary>

```nix
    minecraft-servers = {
      enable = true;
      eula = true;
      dataDir = "/var/lib/minecraft";
      servers.smp = rec {
        enable = true;
        restart = "always";
        package = inputs.nix-minecraft.legacyPackages.${pkgs.system}.minecraftServers.fabric-1_21_4;
        jvmOpts = "-Xms8G -Xmx8G";
        serverProperties = {
          white-list = true;
          difficulty = 3; # hard
          gamemode = 0; # survival
          motd = "Powered by ${package}";
          hide-online-players = true;
          spawn-protection = 0;
          enable-rcon = false; # remote access to server console
        };
        files = {
          "ops.json" = pkgs.writeText "ops.json" (
            builtins.toJSON [
              {
                uuid = "81f9f3cb-646d-4b6b-98a7-800893714036";
                name = "Jappie3";
                level = 4;
              }
            ]
          );
        };
        whitelist = {
          jappie3 = "81f9f3cb-646d-4b6b-98a7-800893714036";
        };
        symlinks = {
          mods = pkgs.linkFarmFromDrvs "mods" (
            builtins.attrValues {
              FabricAPI = pkgs.fetchurl {
                # common hooks and intercompatibility measures for mods using the Fabric toolchain
                url = "https://cdn.modrinth.com/data/P7dR8mSH/versions/15ijyoD6/fabric-api-0.113.0%2B1.21.4.jar";
                sha512 = "sha512-QvY8Lif+1mQJ2Qem7iiUb0yqq/LGRD6vyKEV4S5CeMzXIYvGWTSVrboxEdyayXcQIQeXYIwqiOKnFFP8NAlqmQ==";
              };
              Lithium = pkgs.fetchurl {
                # game logic optimization mod
                url = "https://cdn.modrinth.com/data/gvQqBUqZ/versions/t1FlWYl9/lithium-fabric-0.14.3%2Bmc1.21.4.jar";
                sha512 = "sha512-Gk6vvc7jiG0zwEqkYtE6jB40X/SSABrdJiR2WFt4MnotAW5WOFvO2GlhW8lxYaNKCnFvX1ecjB1wgLJ49PERgw==";
              };
              FerriteCore = pkgs.fetchurl {
                # memory usage optimizations
                url = "https://cdn.modrinth.com/data/uXXizFIs/versions/IPM0JlHd/ferritecore-7.1.1-fabric.jar";
                sha512 = "sha512-9B3J6LKDJ6HimxRmfLQq5efhe8+kSVJg9vhRqA1LCNmKMNXFKxEAB+4yXwLax0MeP61FYMaECvC/NHr61IxarA==";
              };
              Chunky = pkgs.fetchurl {
                # pre-generates chucks
                url = "https://cdn.modrinth.com/data/fALzjamp/versions/VkAgASL1/Chunky-Fabric-1.4.27.jar";
                sha512 = "sha512-qJ+UlH58OZLgHka+iWfSplkzNDM6VGtP/1/bAqH1prg8k63Exyqbmx8U+Sme/KqKXX9e7t89pUHH5yq8Xickxg==";
              };
              CarpetMod = pkgs.fetchurl {
                # take full control of what matters from a technical perspective of the game
                url = "https://cdn.modrinth.com/data/TQTTVgYE/versions/ZF8ufR9V/fabric-carpet-1.21.2-1.4.158%2Bv241022.jar";
                sha512 = "sha512-X6zk5qbbgJEPGbrb2zX95M4l8G/NTi3s5wBc7UygahBgU6Av1Y6mpAqygeuNFe9HtfqUh7Oro18xdI3cNyjSNQ==";
              };
              CarpetTISAddition = pkgs.fetchurl {
                # Carpet mod extension
                url = "https://cdn.modrinth.com/data/jE0SjGuf/versions/zqhgADfG/carpet-tis-addition-v1.65.0-mc1.21.4.jar";
                sha512 = "sha512-/eHdrfR5VX6rm6WTVlMvobaD8xDQ+qpMeAV2p18RTa4sm5rhcOISGa9iCqysvisjeYygRB7IjGBNyiFsp0DA6Q==";
              };
              ChunkDebug = pkgs.fetchurl {
                # tool for debugging chunk loading
                url = "https://cdn.modrinth.com/data/zQxjhDPq/versions/2k4pPXKF/ChunkDebug-2.1.3%2B1.21.4.jar";
                sha512 = "sha512-2pQ8X1vLjstUekBw2QOns4U3fyDfo/J1WMBoJSIwOlJzL8IgURMwXAw+MIsC4rYJw+ygiB6DbcyTE5Ip9tqINg==";
              };
              NoChatReports = pkgs.fetchurl {
                # strips cryptographic signatures from messages because I don't need microsoft 'moderating' *my* server running on *my* hardware
                url = "https://cdn.modrinth.com/data/qQyHxfxd/versions/9xt05630/NoChatReports-FABRIC-1.21.4-v2.11.0.jar";
                sha512 = "sha512-00OwXI5Q8d4VeR/2Iq1E7sps3LIelgomehfXFQbGHKebHIJBZ3eeRNd4yhjcvevllP8jT741W2jSXNtbav1uTw==";
              };
            }
          );
        };
      };
    };
```

</details>

**Comment by Jappie3** [comment]
nvm that - it passes `MAINPID` because the systemd socket option introduced 3 months ago uses it: https://github.com/Infinidoge/nix-minecraft/blob/1ea4cbc1f160175bea2c45a74c4bba048cf071ef/modules/minecraft-servers.nix#L175

maybe add an if statement so to only pass `MAINPID` if `ms.systemd-socket.enable` is enabled?

**Comment by Jappie3** [comment]
so uh
turns out that this was caused by incompatibilities by mods?

I set the managementSystem to use a systemd socket:

```nix
minecraft-servers.managementSystem.systemd-socket.enable = true;
```

after which I saw these logs:

```
Dec 29 10:28:34 Flugel systemd[1]: Starting Minecraft Server smp...
Dec 29 10:28:35 Flugel systemd[1]: Started Minecraft Server smp.
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: Starting net.fabricmc.loader.impl.game.minecraft.BundlerClassPathCapture
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: 2024-12-29T09:28:35.686235661Z main WARN The use of package scanning to locate Log4j plugins is deprecated.
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: Please remove the `packages` attribute from your configuration file.
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: See https://logging.apache.org/log4j/2.x/faq.html#package-scanning for details.
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/INFO]: Loading Minecraft 1.21.4 with Fabric Loader 0.16.9
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/WARN]: Mod resolution failed
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/INFO]: Immediate reason: [HARD_DEP_NO_CANDIDATE carpet-tis-addition 1.65.0 {depends carpet @ [>=1.4.161]}, ROOT_FORCELOAD_SINGLE carpet-tis-addition 1.65.0]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/INFO]: Reason: [HARD_DEP carpet-tis-addition 1.65.0 {depends carpet @ [>=1.4.161]}]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/INFO]: SERVER environment disabled: [fabric-model-loading-api-v1, fabric-renderer-api-v1, fabric-keybindings-v0, fabric-client-tags-api-v1, fabric-sound-api-v1, fabric-screen-api-v1, fabric-key-binding-api-v1, fabric-rendering-v1, fabric-renderer-indigo, fabric-blockrenderlayer-v1]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/INFO]: Fix: add [], remove [], replace [[carpet 1.4.158+v241022] -> add:carpet 1.4.161 ([[1.4.161,∞)])]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: [10:28:35] [main/ERROR]: Incompatible mods found!
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: net.fabricmc.loader.impl.FormattedException: Some of your mods are incompatible with the game or each other!
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: A potential solution has been determined, this may resolve your problem:
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:          - Replace mod 'Carpet Mod' (carpet) 1.4.158+v241022 with version 1.4.161 or later.
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]: More details:
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:          - Mod 'Carpet TIS Addition' (carpet-tis-addition) 1.65.0 requires version 1.4.161 or later of mod 'Carpet Mod' (carpet), but only the wrong version is present: 1.4.158+v241022!
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:         at net.fabricmc.loader.impl.FormattedException.ofLocalized(FormattedException.java:51) ~[n7xayh1wvphviyyba254gc7djisfml5i-fabric-server-launch.jar-fabric-0.16.9-1.21.4:9.7.1]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:         at net.fabricmc.loader.impl.FabricLoaderImpl.load(FabricLoaderImpl.java:196) ~[n7xayh1wvphviyyba254gc7djisfml5i-fabric-server-launch.jar-fabric-0.16.9-1.21.4:9.7.1]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:         at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:146) ~[n7xayh1wvphviyyba254gc7djisfml5i-fabric-server-launch.jar-fabric-0.16.9-1.21.4:9.7.1]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:         at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:68) [n7xayh1wvphviyyba254gc7djisfml5i-fabric-server-launch.jar-fabric-0.16.9-1.21.4:9.7.1]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:         at net.fabricmc.loader.impl.launch.knot.KnotServer.main(KnotServer.java:23) [n7xayh1wvphviyyba254gc7djisfml5i-fabric-server-launch.jar-fabric-0.16.9-1.21.4:9.7.1]
Dec 29 10:28:35 Flugel minecraft-server-smp-start[213226]:         at net.fabricmc.loader.impl.launch.server.FabricServerLauncher.main(FabricServerLauncher.java:69) [n7xayh1wvphviyyba254gc7djisfml5i-fabric-server-launch.jar-fabric-0.16.9-1.21.4:9.7.1]
Dec 29 10:28:35 Flugel systemd[1]: minecraft-server-smp.service: Main process exited, code=exited, status=1/FAILURE
Dec 29 10:28:36 Flugel systemd[1]: minecraft-server-smp.service: Failed with result 'exit-code'.
Dec 29 10:28:36 Flugel systemd[1]: Stopped Minecraft Server smp.
Dec 29 10:28:36 Flugel systemd[1]: minecraft-server-smp.service: Consumed 2.446s CPU time, 132.2M memory peak, 4K written to disk.
```

Updating Carpetmod to a more recent version fixed the problem.
